### PR TITLE
[SPARK-40633][BUILD] Upgrade janino to 3.1.9

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr aims upgrade janino from 3.1.7 to 3.1.9

### Why are the changes needed?
This version bring some improvement and bug fix, and janino 3.1.9 will no longer test Java 12, 15, 16 because these STS versions have been EOL:

- https://github.com/janino-compiler/janino/compare/v3.1.7...v3.1.9

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Pass GitHub Actions
- Manual test this pr with Scala 2.13, all test passed

### CODE REVIEW

**Comment by LuciferYang** [comment]
Test first

**Comment by LuciferYang** [comment]
Some test failed, will investigate later

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 1]
```diff
@@ -1310,7 +1310,7 @@ case class CatalystToExternalMap private(
 
     val tupleClass = classOf[(_, _)].getName
     val appendToBuilder = s"""
-      $tupleClass $tupleLoopValue;
+      $tupleClass $tupleLoopValue = null;
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:1313]: init to null for fix

```
/* 063 */         scala.Tuple2 tupleLoopValue_0;
/* 064 */
/* 065 */         if (false) {
/* 066 */           tupleLoopValue_0 = new scala.Tuple2(value_CatalystToExternalMap_key_lambda_variable_1, null);
/* 067 */         } else {
/* 068 */           tupleLoopValue_0 = new scala.Tuple2(value_CatalystToExternalMap_key_lambda_variable_1, value_CatalystToExternalMap_value_lambda_variable_2);
/* 069 */         }
/* 070 */
/* 071 */         builderValue_0.$plus$eq(tupleLoopValue_0);
/* 072 */
```

```
Caused by: org.codehaus.commons.compiler.InternalCompilerException: File 'generated.java', Line 71, Column 16: Compiling "builderValue_0.$plus$eq(tupleLoopValue_0)"
	at org.codehaus.janino.UnitCompiler.compileGetValue(UnitCompiler.java:5819)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:4053)
	at org.codehaus.janino.UnitCompiler.access$6100(UnitCompiler.java:236)
	at org.codehaus.janino.UnitCompiler$13.visitMethodInvocation(UnitCompiler.java:4028)
	at org.codehaus.janino.UnitCompiler$13.visitMethodInvocation(UnitCompiler.java:4003)
	at org.codehaus.janino.Java$MethodInvocation.accept(Java.java:5470)
	at org.codehaus.janino.UnitCompiler.compile(UnitCompiler.java:4003)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:2487)
	... 155 more
Caused by: org.codehaus.commons.compiler.InternalCompilerException: File 'generated.java', Line 71, Column 25: Compiling "tupleLoopValue_0"
	at org.codehaus.janino.UnitCompiler.compileGetValue(UnitCompiler.java:5819)
	at org.codehaus.janino.UnitCompiler.compileGet2(UnitCompiler.java:5323)
	at org.codehaus.janino.UnitCompiler.access$9300(UnitCompiler.java:236)
	at org.codehaus.janino.UnitCompiler$16.visitMethodInvocation(UnitCompiler.java:4698)
	at org.codehaus.janino.UnitCompiler$16.visitMethodInvocation(UnitCompiler.java:4674)
	at org.codehaus.janino.Java$MethodInvocation.accept(Java.java:5470)
	at org.codehaus.janino.UnitCompiler.compileGet(UnitCompiler.java:4674)
	at org.codehaus.janino.UnitCompiler.compileGetValue(UnitCompiler.java:5817)
	... 162 more
Caused by: org.codehaus.commons.compiler.InternalCompilerException: Invalid local variable index 10
	at org.codehaus.janino.UnitCompiler.getLocalVariableTypeInfo(UnitCompiler.java:13498)
	at org.codehaus.janino.UnitCompiler.load(UnitCompiler.java:12500)
	at org.codehaus.janino.UnitCompiler.load(UnitCompiler.java:12475)
	at org.codehaus.janino.UnitCompiler.compileGet2(UnitCompiler.java:4745)
	at org.codehaus.janino.UnitCompiler.access$8200(UnitCompiler.java:236)
	at org.codehaus.janino.UnitCompiler$16$1.visitLocalVariableAccess(UnitCompiler.java:4684)
	at org.codehaus.janino.UnitCompiler$16$1.visitLocalVariableAccess(UnitCompiler.java:4678)
	at org.codehaus.janino.Java$LocalVariableAccess.accept(Java.java:4661)
	at org.codehaus.janino.UnitCompiler$16.visitLvalue(UnitCompiler.java:4678)
	at org.codehaus.janino.UnitCompiler$16.visitLvalue(UnitCompiler.java:4674)
	at org.codehaus.janino.Java$Lvalue.accept(Java.java:4528)
	at org.codehaus.janino.UnitCompiler.compileGet(UnitCompiler.java:4674)
	at org.codehaus.janino.UnitCompiler.compileGet2(UnitCompiler.java:4741)
	at org.codehaus.janino.UnitCompiler.access$7700(UnitCompiler.java:236)
	at org.codehaus.janino.UnitCompiler$16$1.visitAmbiguousName(UnitCompiler.java:4679)
	at org.codehaus.janino.UnitCompiler$16$1.visitAmbiguousName(UnitCompiler.java:4678)
	at org.codehaus.janino.Java$AmbiguousName.accept(Java.java:4603)
	at org.codehaus.janino.UnitCompiler$16.visitLvalue(UnitCompiler.java:4678)
	at org.codehaus.janino.UnitCompiler$16.visitLvalue(UnitCompiler.java:4674)
	at org.codehaus.janino.Java$Lvalue.accept(Java.java:4528)
	at org.codehaus.janino.UnitCompiler.compileGet(UnitCompiler.java:4674)
	at org.codehaus.janino.UnitCompiler.compileGetValue(UnitCompiler.java:5817)
	... 169 more
```

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:1313]: The root cause of the change has not been investigated

**Comment by LuciferYang** [comment]
will check 3.1.9

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 1 above)
Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: 3.1.9 should fix this issue, upgrade to check

**Comment by LuciferYang** [comment]
3.1.9 fixed the compatibility issue of 3.1.8, GA passed

**Comment by LuciferYang** [comment]
cc @HyukjinKwon @srowen

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by cloud-fan** [comment]
This is a very critical dependency for Spark SQL. Can we run a TPCDS benchmark before upgrading?

**Comment by LuciferYang** [comment]
> 

sf-1 use GA?

**Comment by LuciferYang** [comment]
@cloud-fan Have you found any bad cases now?

**Comment by LuciferYang** [comment]
I will give a followup to update the tpcds result, then we can choose whether to keep the version or revert it

**Comment by cloud-fan** [comment]
We are going to evaluate, but it's better if the PR author can do some local benchmarks to verify the performance first, as big regressions can be easily observed locally.

**Comment by LuciferYang** [comment]
> We are going to evaluate, but it's better if the PR author can do some local benchmarks to verify the performance first, as big regressions can be easily observed locally.

The suggestion is good. Very sorry for the preparation work before this upgrade was not enough

**Comment by LuciferYang** [comment]
@cloud-fan 

- Before this pr merge([SPARK-40834][SQL][FOLLOWUP] Take care of legacy query end events) use 3.1.7
[TPCDSQueryBenchmark-results.txt](https://github.com/apache/spark/files/10123191/TPCDSQueryBenchmark-results.txt)

**Sum(Avg) is  479642ms**

- After this pr merge use 3.1.9
[TPCDSQueryBenchmark-results.txt](https://github.com/apache/spark/files/10123193/TPCDSQueryBenchmark-results.txt)

**Sum(Avg) is 475230ms**

Seems no obvious change.

**Pull request closed without merging**